RFD 0086: Prevent user role self lock (#15706)
gravitational/teleport: rfd/0083-prevent-user-role-self-lock.md

>
> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
> 4. SSO users can have full access.

The solution here could be to introduce a new rule:

- There should be at least _admin_

This will ensure that:

- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
- Locking will be impossible on the SSO side since there will be always _admin_ on our side.

Now we need to define what _admin_ means in our system. W can take 2 approaches:

1. User with an immutable `editor` role is _admin_.

> 2. User with a role that grants user/role editing capabilities is _admin_.

Author: > Please correct me if I am wrong, isn't role editing capabilities (role with create/update rules to the role resource) just enough?

If there is only one user that can edit/create/delete new roles (no other user), he cannot create a new user with a role or assign a role to some other user. He is locked. So the _admin_ should have a combination of user/role editing capabilities

> 2. User with a role that grants user/role editing capabilities is _admin_.

Author: > If an admin lost user editing capabilities while retaining role editing capabilities, they can still edit the role and add back those user editing capabilities
Right, because he still has at least one role which he can edit 👍🏻

From the user's perspective, the first approach is simpler. The error message saying: "You can't delete the last user with `editor` role." will be easier to understand than "You can't delete the last user with: list of full user & role management capabilities".

Also, implementation would be slightly easier. We wouldn't need to define validation for editing roles (since the `editor` role would be immutable).

The drawback here is that we force users to have the `editor` role.

### UI and behavior changes

> We need to change the UI to reflect the new policy. We need to add validation to all actions that can cause lock (in `WebUI` and `tctl`). The error messages should inform the user why this action is not allowed.

Author: Yes, exactly. We need a central place for making decision and respond to both clients and this will be our backend

Teleport already protects against a similar case in the following scenario:

| action       | initial state | change                           |
| ------------ | ------------- | -------------------------------- |
| role removal | _admin_       | _admin_ tries to remove own role |

The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.

There is one more case when a user can lock himself out:

| initial state | change                                                                                    |
| ------------- | ----------------------------------------------------------------------------------------- |
| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |

However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.

## Terminology

- _admin_: local user with user/role editing capabilities
- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)

## Why

> When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.

Reviewer: ```suggestion
When a user locks themselves out, they are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
```

> Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.

Author: No, but I think @pierrebeaucamp will know more

> Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.

Reviewer 2: Thanks for the answer, folks. Makes sense.

Jan, I think it's worth mentioning in the RFD that we don't do account resets because of potential social engineering attacks, and that Cloud users are expected to rely on their recovery codes instead.

> On-prem customers could work around this "locked out" state by manually instrumenting the auth process using tctl.

Reviewer: ```suggestion
On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
```

> ## Details

Reviewer 2: Some high-level feedback:

1. I think we need to discuss cluster types separately: on-prem local, on-prem SSO and Cloud. Of all of those, it seems that Cloud is the one that wants this feature the most - on-prem clusters can be accessed by other ways, meaning that `tctl` interventions are possible.

2. Given that Cloud is the one that wants this the most, let's talk about the "no reset" policy (and why we are still keeping it).

3. This is minor, but I find it better to have a design that states its preferred solution directly, moving alternatives to an "Alternatives Considered" section - that makes it clear what is the actual design vs what we considered (but didn't do).

> Context: SSO users
>
> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
>
> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
> 4. SSO users can have full access.

The solution here could be to introduce a new rule:

> - There should be at least _admin_

Reviewer 2: Do you mean that "There should be at least one _admin_ user" ? The sentence seems incomplete.

Author: This describes a state where some user is able to manage cluster

This will ensure that:

- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
- Locking will be impossible on the SSO side since there will be always _admin_ on our side.

Now we need to define what _admin_ means in our system. W can take 2 approaches:

> 1. User with an immutable `editor` role is _admin_.

Reviewer 2: Have we considered having a magical "admin"-like user that can never be deleted, akin to a Linux root?

Given that it's likely a common idea, I think it's worth having a small section in the design comparing our solution to it and saying why we didn't go this way.

Author: Yeah, I have considered it. The problem is that customers would need to add this root user to each cluster.

Reviewer 2: That sounds doable, though.

I wish we registered a range of user names for system use. What's done is done, but maybe we could take advantage of your RFD and introduce that rule for 11+ clusters.

Author: @xinding33 What do you think?

> 2. User with a role that grants user/role editing capabilities is _admin_.

Reviewer: Are we only considering local users here? It seems that way from the implementation PR, should make that clear if so.

It seems that with this design we will be explicitly forcing the existence of a local user with role edit permissions? I could see users wanting to avoid this so that everything MUST go through SSO.

Many existing clusters will not have this property when upgrading, and newly created clusters have zero users so they will also not satisfy this property. We should be careful about what to do in this scenario so that SSO admin users and local `tctl` are not blocked from editing roles (or any other actions)

Author: > Are we only considering local users here?

Taking only local users into consideration will ensure that in case of SSO users deletion / SSO disconnection there will be still someone able to restore the cluster.

> Many existing clusters will not have this property when upgrading, and newly created clusters have zero users so they will also not satisfy this property. We should be careful about what to do in this scenario so that SSO admin users and local tctl are not blocked from editing roles (or any other actions)

Good point! We need to consider this. I think is safe to assume that currently there are clusters with only SSO users and they don't want local users. 
I think there can be two approaches here:
1. treat the SSO users as regular users. This will allow lock situations when destructive changes are made on the SSO provider side (disconnect, user deletion, group deletion, etc...). We can't control them. We need to assume that 
if a connector is registered and grants access to a role that allows managing the cluster this means we are also in a valid state.

> From the user's perspective, the first approach is simpler. The error message saying: "You can't delete the last user with `editor` role." will be easier to understand than "You can't delete the last user with: list of role management capabilities".

Reviewer 2: +1 for this - simple to explain and to implement. It's easy to have regressions in systems like this, which we are only likely to find out once someone gets locked out (again). It's much simpler to ensure that an immutable role performs whatever actions it needs to.

Author: The immutable roles would require users to modify their cluster: 

> complicated migration in case of updating cluster. It is still possible to delete/edit default roles so not all existing clusters will have default roles in the initial form. Maybe the `editor` role is used as a role with a different meaning than the initial one? We can't override existing roles. So the solution would be to add 3 new roles (instead of editing/re-adding: editor, auditor, access). This will ensure we don't break anything. But the problem will be users would have to assign manually a new _admin_ role to some users.

That's why we decided to proceed with the second option.

Reviewer 2: It looks to me that immutable system roles are something we want to have anyway, _and_ it would make your life simpler, that's why I'm suggesting it here.

We are acting as if it was an unsolvable problem, but it's far from it. For example, we can have an immutable, editor-like system role that uses a fixed UUID as it's ID. It's not pretty, but it's a solution for older clusters. (Reserved role names would be even better, I wish we'd have prepared for that sooner.)

For the purposes of this design we don't need editor/auditor/access, we just need _a_ role that does the job.

Another alternative for this is the superuser approach on thread https://github.com/gravitational/teleport/pull/15706#discussion_r969887140.

Author: The problem is not with the name (we could even generate it) of the role but the need to assign some user to it: 

> But the problem will be users would have to assign manually a new _admin_ role to some users.

The other benefit is that we wouldn't need to define validation for editing roles (since the `editor` role would be immutable).

The drawbacks here are:

- we force users to have the `editor` role
- complicated migration in case of updating cluster. It is still possible to delete/edit default roles so not all existing clusters will have default roles in the initial form. Maybe the `editor` role is used as a role with a different meaning than the initial one? We can't override existing roles. So the solution would be to add 3 new roles (instead of editing/re-adding: editor, auditor, access). This will ensure we don't break anything. But the problem will be users would have to assign manually a new _admin_ role to some users.

> I think `2.` approach is better since migration cost outweighs the benefits of the pros.

Reviewer 2: In practice it seems like option 2. is the same as having a fixed root user in cluster, but with some extra complexity added.

Author: The main difference here is customer impact:
* (fixed root user) Impacts all clusters. Will require manual migration: adding new root user.
* (option 2)  Most existing clusters have users who can edit roles, so there will be no required action. The only exception is when there are only SSO users in the cluster. We can warn them that adding such a user with x role is advisable (after that they cannot go back to only SSO users state).

Reviewer 2: > (fixed root user) Impacts all clusters. Will require manual migration: adding new root user.

Migration, yes, but why does it have to be manual? I don't follow that.

> (option 2) Most existing clusters have users who can edit roles, so there will be no required action (...)

Automatically adding a root-like user solves this too and requires no action on part of users. I do think we should explore this alternative a bit more before discarding it.

Author: Right, we could add a root user automatically using username and password generation (running some procedure when a cluster is updated or initially bootstrapped). 

The authorized user can then reset authentication and set a known password or another form of auth. This is optional but required for the root account to be useful. We could add the warning that it is recommended action.

I think we should consider this: 
I think customers will mostly use passwords as auth since the root user will be not a personal account. Shared passwords are much more likely to be misused, their passwords tend to remain unchanged for extended periods, and often leak when employees change jobs. Also, since they are nobody's personal responsibility and sort fo common knowledge among peers, they tend to not get the same amount of diligence as personal accounts and are often emailed or written down in notes, files, and password managers. Hackers could utilize that to gain access to the cluster.

### UI and behavior changes

> We need to change the UI to reflect the new policy. We need to add validation to all actions that can cause lock (in `WebUI` and `tctl`). The error messages should inform the user why this action is not allowed.

Reviewer 2: We need to change the UI how? It's not clear to me what's in scope.

Author: I meant that we will display dedicated error messages to the user. This doesn't require frontend changes

Reviewer 2: Let's clarify the paragraph, but sounds good.

Author: e295a5985e1d68b2170cda9c622d1d156244c88b

Reviewer 2: Thanks!

---
authors: Jan Kaczmarkiewicz (jan.kaczmarkiewicz@goteleport.com)
state: draft
---

# RFD 83 - Prevent user role self-lock

## Required Approvers

- Engineering: @jimbishopp
- Product: @xinding33

## What

> Users can change the permission in a way that can cause no one to be able to change it further.
> 
> Currently, there are multiple ways for users to lock themselves:

Reviewer 3: ```suggestion
Users can change their permissions in a way that leaves them incapable of making further changes to the cluster - in effect they lock themselves out.

Currently, there are multiple ways to trigger this scenario:
```

> | initial state                              | operation                                                                                   | result |
> | ------------------------------------------ | ------------------------------------------------------------------------------------------- | :----: |
> | 1 _admin_ and _regular users_              | _admin_ removes himself from an `editor` role                                               |   🔒   |
> | 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |   🔒   |
> | 1 _admin_, _SSO admin_ and _regular users_ | _SSO admin_ removes the `editor` role from _admin_ and SSO connector. _SSO admin_ logs out. |   🔒   |
> | _SSO admin_ and _regular users_            | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_      |   🔒   |

Reviewer 3: ```suggestion
| initial state                              | change                                                                                    |
| ------------------------------------------ | ------------------------------------------------------------------------------------------- | :----: |
| 1 _admin_ and _regular users_              | _admin_ removes himself from an `editor` role                                               | 
| 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |   
| 1 _admin_, _SSO admin_ and _regular users_ | _SSO admin_ removes the `editor` role from _admin_ and SSO connector. _SSO admin_ logs out. | 
| _SSO admin_ and _regular users_            | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_      |
```

> Cases handled currently:
> 
> | initial state               | operation                                                                          | result |
> | --------------------------- | ---------------------------------------------------------------------------------- | :----: |
> | _admin_ and _regular users_ | _admin_ tries to delete his account. Operation is not permitted, an error is shown |   🔓   |

Reviewer 3: ```suggestion
Teleport already protects against a similar case in the following scenario:

| initial state               | change                                                                       |
| --------------------------- | ---------------------------------------------------------------------------------- | :----: |
| _admin_ and _regular users_ | _admin_ tries to delete his account. Operation is not permitted, an error is shown | 
```

## Terminology

- _admin_: local user with `editor` role
- _SSO admin_: user with `editor` role added via SSO connector (eg. GitHub)
- _regular user_: user without `editor` role
- 🔒: No one can edit roles. Deadlock

> - 🔓: There is still an active user with an editor role

Reviewer 4: +1

## Why

> When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.

Reviewer 3: Further, for our Cloud offering we don't perform account resets as a policy. The only recurse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data-loss.

On-prem customers could work around this "locked out" state by manually instrumenting the `auth` process using `tctl`.

Author: fec4c734ce2d1180ca2f0b5715533d90cc5b0fcd

## Details

> Context: SSO users
>

> > Teleport has this feature called SSO auth. It allows to addition SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:

Reviewer 3: ```suggestion
> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
```

>
> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
> 4. SSO users can be `editors`.

The solution here could be to introduce a new rule:

- There should be at least 2 local _admins_

This will ensure that:

- When one of the _admins_ loses access to the account the second one will operate.
- Users won't be able to delete _admin_ user if there are only 2 _admins_. They would have to add a third one and then they can delete it.
- Users won't be able to unassign the `editor` role from _admin_ user if there are only 2 _admins_. They would have to add a third and then they can unassign.
- Role system lock will be independent of connector removal or SSO changes: team deletion/user deletion.

Should we enforce this rule when a user is setting up its cluster? What about existing users? I think there are two approaches here:

> - they would have to create immediately a second _admin_ user (if they have only one currently). They cannot use a cluster if they don't.

Reviewer 3: I don't think this is a viable option. Especially for trials or POCs, there might only be a single user interacting with the system. This would also impact any OSS users that might run Teleport for their personal access needs.

Reviewer 5: That's fair. I agree, I think we should explore Naji's proposed approach (https://github.com/gravitational/teleport/pull/15706/files#r953966589) instead of mandating 2 admin users.

---
authors: Jan Kaczmarkiewicz (jan.kaczmarkiewicz@goteleport.com)
state: draft
---

# RFD 83 - Prevent user role self-lock

## Required Approvers

- Engineering: @jimbishopp
- Product: @xinding33

## What

Users can change their permissions in a way that leaves them incapable of making further changes to the cluster - in effect they lock themselves out.

Currently, there are multiple ways to trigger this scenario:

| initial state                              | change                                                                                      |
| ------------------------------------------ | ------------------------------------------------------------------------------------------- |
| 1 _admin_ and _regular users_              | _admin_ removes himself from an `editor` role                                               |

> | 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |

Author: Yeah, but still this leaves space for user error. eg. locally stored recovery codes & generated passwords in the password manager will get lost if the computer disk goes down. The cluster will be lost. I mean it is unlikely but possible

- _regular user_: user without `editor` role

## Why

When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.

Further, for our Cloud offering we don't perform account resets as a policy. The only recurse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data-loss.

On-prem customers could work around this "locked out" state by manually instrumenting the auth process using tctl.

## Details

> Context: SSO users
>
> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
>
> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
> 4. SSO users can be `editors`.

> The solution here could be to introduce a new rule:

Author: > As Pierre mentioned, I think the two admins solution is not really necessary.

The requirement for a second admin guards cluster from situations where there is one admin and he loses the account.

> The solution here could be to introduce a new rule:

Author: If we are ok with this case, this will greatly simplify rfd & implementation.

> The solution here could be to introduce a new rule:

Reviewer 5: Let's explore if @NajiObeid 's suggestion [here](https://github.com/gravitational/teleport/pull/15706/files#r953966589) would address this lockout case.

- There should be at least 2 local _admins_

> This will ensure that:

Reviewer: > preventing the deletion the last role with create/update rules to the role resource.

We already prevent deletion of any role which is currently assigned to a user, I don't think we should need any extra check for this.

> preventing the removal of create/update rules from the last role that's got them.

To be more precise, if there is role `admin` and `admin-nobody-has-this-role`, and both have create/update role rules, we still need to prevent the removal of create/update rules from `admin`. The other role with the requisite rules still exists, but there is no user who is assigned it

- When one of the _admins_ loses access to the account the second one will operate.
- Users won't be able to delete _admin_ user if there are only 2 _admins_. They would have to add a third one and then they can delete it.

> - Users won't be able to unassign the `editor` role from _admin_ user if there are only 2 _admins_. They would have to add a third and then they can unassign.

Reviewer 5: I agree that ideally we shouldn't force specific roles on users. But I think it comes down to implementation complexity and user experience. If it's an order of magnitude less work to enforce at least one user with the editor role, then I think we should go with that implementation for now and we can always improve it in the future. Regarding user experience, it's very easy for users to understand "immutable default roles + at least 1 editor" but it's a lot harder for them to understand "you need this specific role option in at least one role assigned to you..."
